src/reference/mod.rs

```diff
@@ -0,0 +1,367 @@
+//! regular expressions matches.  If more than one regular expression matches at this earliest
+//! character, we use the regular expression that appears first in the `scan` statement.  We then
+//! execute the statements in the "winning" regular expression's block.
+//!
+//! After executing the matching block, we try to match all of the regular expressions again,
+//! starting _after_ the text that was just matched.  We continue this process, applying the
+//! earliest matching regular expression in each iteration, until we have exhausted the entire
+//! string, or none of the regular expressions match.
+//!
+//! Within each regular expression's block, you can use `$0`, `$1`, etc., to refer to any capture
+//! groups in the regular expression.
+//!
+//! For example, if `filepath` is a global variable containing the path of a Python source file,
+//! you could use the following `scan` statement to construct graph nodes for the name of the
+//! module defined in the file:
+//!
+//! ``` tsg
+//! (module) @mod
+//! {
+//!   var current = @mod.root
+//!
+//!   scan filepath {
+//!     "([^/]+)/"
+//!     {
+//!       ; This arm will match any directory component of the file path.  In
+//!       ; Python, this gives you the sequence of packages that the module
+//!       ; lives in.
+//!
+//!       ; Note that we keep appending additional tag names to the `current`
+//!       ; graph node to create an arbitrary number of graph nodes linked to
+//!       ; the @mod syntax node.
+//!       set current = current.package
+//!       attr current "name" = $1
+//!     }
+//!
+//!     "__init__\\.py$"
+//!     {
+//!       ; This arm will match a trailing __init__.py, indicating that the
+//!       ; module's name comes from the last directory component.
+//!
+//!       ; Expose the graph node that we created for that component as a
+//!       ; scoped variable that later stanzas can see.
+//!       let @mod::root = current
+//!     }
+//!
+//!     "([^/]+)\\.py$"
+//!     {
+//!       ; This arm will any other trailing module name.  Note that
+//!       ; __init__.py also matches this regular expression, but since it
+//!       ; appears later, the __init__.py clause will take precedence.
+//!
+//!       set current = current.package
+//!       attr current "name" = $1
+//!       let @mod::root = current
+//!     }
+//!   }
+//! }
+//! ```
+//!
+//! # Functions
+//!
+//! The process executing a graph DSL file can provide **_functions_** that can be called from
+//! within graph DSL stanzas.
+//!
+//! Functions calls use a Lisp-like syntax, where the name of the function being called is _inside_
+//! of the parentheses.  The parameters to a function call are arbitrary expressions.  For
+//! instance, if the executing process provides a function named `+`, you could call it as:
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!    let x = 4
+//!    attr @id.node "nine" = (+ x 5)
```

Comment:
Does this imply that each syntax node / tag pair has distinct attributes?

For example:

```shell
(identiifer) @id
{
  let x = 4
  attr @id.node1 "nine" = (+ x 5)
  attr @id.node2 "ten" = (+ x 6)
}
```

Then when referencing the `@id` graph node, there would be two tags, and each tag would have one attribute associated with it?

Reply:
`@id` isn't a graph node, it's a syntax node.  `@id.node1` and `@id.node2` are two distinct graph nodes, both "attached" to the `@id` syntax node, and each graph node has one attribute.

The whole description of syntax / graph nodes could use some rewording.

Reply:
Technically `@id` is also a graph node, since the list of tags can be empty. It's not uncommon to do stuff like `edge @foo -> @foo.bar`, at least insofar as anything can be common or uncommon currently.